### PR TITLE
I14 service layer

### PR DESCRIPTION
**Pull request by peter-gy**
Implements #14.

To handle the different query arguments in a generic way, I introduced an abstract class-based approach in [ufind_queries.dart](https://github.com/memoriesadrift/we-find/blob/i14-service-layer/lib/service/ufind_queries.dart). It makes it possible to access the different query options through the same interface and at the same time it is very easy to add support for further query options in the future.

Currently the following options are implemented:
- Set the number of search results
- Exact, non-fuzzy search
- Filter for courses with streaming
- Search for courses of a certain directorate of studies
- Search for courses in certain semesters
- Search for courses in the current semester
- Search for courses in the current study year
- Sort search results alphabetically
- Sort courses by course number

The implementation is fully compliant with the specifications of the [u:find help page](https://ufind.univie.ac.at/en/help.html)

Unfortunately, it seems like the API does not support chaining multiple query arguments. This seems pretty unrealistic though and I hope that I am only missing something very obvious. For the former case, the function [queryForCourses](https://github.com/memoriesadrift/we-find/blob/i14-service-layer/lib/service/ufind_queries.dart#L47) supports only one `QueryArgument`. If it turns out that the latter case is true, the code can be easily modified to handle chained arguments.

### CODE REVIEW

File: lib/service/ufind_service.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+const authority = "m1-ufind.univie.ac.at";
+
+/// Returns the root [StudyModule] based on the specified semester [when]
+/// and the [spl].
+///
+/// Calling [fetchStudyModule('2021S', 5)] would return the root [StudyModule]
+/// of the Computer Science faculty of the summer semester 2021.
+/// Upon providing a [when] value that is in the future, such as 2030S,
+/// the request will be successful, however the returned [StudyModule] will be null.
+Future<StudyModule?> fetchStudyModule(String when, int spl) {
```
Comment (michalrzak) [inline lib/service/ufind_service.dart:18]: Is every module accessible by spl? How would we for example fetch ```5.02 Bachelor Computer Science 2016 (521 [4]) - Data Science```

Comment (peter-gy) [reply lib/service/ufind_service.dart:18]: `5.02 Bachelor Computer Science 2016 (521 [4]) - Data Science` will be a direct submodule of the module that gets returned by this call. The `List<StudyModule> submodules` instance of the return value of `fetchStudyModule('2021S', 5)` will contain the desired module.

File: lib/service/ufind_service.dart [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+import 'package:we_find/extensions/string_extensions.dart';
+import 'package:we_find/extensions/xml_extensions.dart';
+import 'package:we_find/model/model.dart';
+
+const authority = "m1-ufind.univie.ac.at";
+
+/// Returns the root [StudyModule] based on the specified semester [when]
```
Comment (memoriesadrift) [inline lib/service/ufind_service.dart:11]: Nitpick, but let's rename `when` to something else. `when` sounds like a reserved keyword. Let's use `semester` or `moduleDate` or something similar.

**Review by memoriesadrift — APPROVED** [review]

File: lib/service/ufind_service.dart (hunk 1 above)
Comment (michalrzak) [reply lib/service/ufind_service.dart:18]: Seems workable. If it is turns out tedious or something later we can change that then.

**Review by michalrzak — APPROVED** [review]

**Pull request merged**